kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,952 @@
+			case api.KindIPPool:
+				c.handleIPPoolUpdate(update)
+				kick(c.syncChan)
+				return
+			}
+		case model.BlockKey:
+			c.handleBlockUpdate(update)
+			return
+		}
+	}
+}
+
+func (c *loadBalancerController) handleBlockUpdate(kvp model.KVPair) {
+	if kvp.Value != nil {
+		affinity := kvp.Value.(*model.AllocationBlock).Affinity
+
+		if affinity != nil && *affinity == fmt.Sprintf("%s:%s", ipam.AffinityTypeVirtual, api.VirtualLoadBalancer) {
+			c.ipamBlocks[kvp.Key.String()] = kvp
+		}
+	} else {
+		delete(c.ipamBlocks, kvp.Key.String())
+	}
+}
+
+func (c *loadBalancerController) handleIPPoolUpdate(kvp model.KVPair) {
+	if kvp.Value != nil {
+		pool := kvp.Value.(*api.IPPool)
+		if slices.Contains(pool.Spec.AllowedUses, api.IPPoolAllowedUseLoadBalancer) {
+			c.ipPools[kvp.Key.String()] = *pool
+		}
+	} else {
+		delete(c.ipPools, kvp.Key.String())
+	}
+}
+
+// syncIpam has two main uses. It functions as a garbage collection for leaked IP addresses from Service LoadBalancer
+// The other use case is to update IPs for any Service LoadBalancer that do not have IPs assigned, this could be caused by the user
+// creating Service LoadBalancer before any valid pools were created
+func (c *loadBalancerController) syncIPAM() {
+	svcObjs, err := c.getServiceObjectList()
+	if err != nil {
+		return
+	}
+
+	for _, svcObj := range svcObjs {
+		err = c.syncService(svcObj)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to sync service %s/%s", svcObj.service.Namespace, svcObj.service.Name)
+			continue
+		}
+	}
+
+	for _, block := range c.ipamBlocks {
+		b := block.Value.(*model.AllocationBlock)
+		for key := range b.SequenceNumberForAllocation {
+			ordinal, err := strconv.Atoi(key)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to parse Ordinal for block %s", key)
+			}
+			ip := b.OrdinalToIP(ordinal)
+
+			if _, ok := c.allocationTracker.servicesByIp[ip.String()]; !ok {
+				log.Infof("Found allocated IP, but not in use. Will release IP: %s", ip.String())
+				releaseOptions := ipam.ReleaseOptions{
+					Address: ip.String(),
+				}
+				_, err := c.calicoClient.IPAM().ReleaseIPs(context.Background(), releaseOptions)
+				if err != nil {
+					log.WithError(err).Errorf("Failed to release IP %s", ip.String())
+					continue
+				}
+			}
+		}
+	}
+}
+
+// syncService does the following:
+// - Releases any IP addresses in the IPAM DB associated with the Service that are not in the Service status.
+// - Allocates any addresses necessary to satisfy the Service LB request
+// - Updates the controllers internal state tracking of which IP addresses are allocated.
+// - Updates the IP addresses in the Service Status to match the IPAM DB.
+func (c *loadBalancerController) syncService(svcObj serviceObject) error {
+	var err error
+	var svc *v1.Service
+	serviceUpdated := false
+	reassignIP := false
+	svcStatus := make(map[string]v1.LoadBalancerIngress)
+
+	switch svcObj.updateType {
+	case serviceUpdateTypeDELETE:
```

Comment:
I still dislike our sync loop being aware of event types - it shouldn't care about whether we received a DELETE event or not. Instead, it should care about whether or not the service actually exists.

I would prefer if this function took a `types.NamespacedName` and used it to look for the service object in the cache.

```go
func (c *loadBalancerController) syncService(key types.NamespacedName) error {

  // Check if the service exists.
  svc, err := c.serviceLister.Services(key.Namespace).Get(key.Name)
  if err != nil { . . . }

  if svc == nil {
    // Service doesn't exist - clean up state and release any IPs
  } else {
    // Service exists - sync it.
  }
}
```


I don't think this changes the logic you have written here in any substantial way, except:
- It simplifies the `serviceObject` type - it's only used on the channel to signal events. In fact, we could remove the type entirely and just send a `NamespacedName` (it doesn't need to specify an event type, because we can use the presence / absence of the service in the cache to determine what course of action to take).
- It correctly signals that this code is not event driven, rather it's operating based on cluster state.